inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,190 +1041,541 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
+    } else if (mayiuse(cpu::x64::sse41)) {
+        blk_size = 8;
+        count_xmm = 16;
+    }
+
+    // [case 1]: if 2 * (top_k + 1) + 2 <= count_vec, thus top_k is relatively small, and vector registers are sufficient
+    //           to keep all necessary data for sorting, no need to load and store frequently, use bubble sort;
+    // [case 2]: otherwise, only when topk is imposed on innermost dimsension of planar(ncsp/nspc) layout, should heap sort be used;
+    // [case 3]: bitonic sort is used as default
+    if (top_k <= count_xmm / 2 - 2) {
+        algorithm = TopKAlgorithm::topk_bubble_sort;
+    } else if ((layout == TopKLayoutType::topk_ncsp || layout == TopKLayoutType::topk_nspc) && topk_innermost) {
+        algorithm = TopKAlgorithm::topk_heap_sort;
+    } else {
+        algorithm = TopKAlgorithm::topk_bitonic_sort;
+    }
+
+    if (jit_mode) {
+        O = 1, I = 1;
+        A = src_dims[axis];
+        if (layout == TopKLayoutType::topk_ncsp) {
+            for (int i = 0; i < axis; i++)
+                O *= src_dims[i];
+            for (size_t i = axis + 1; i < src_dims.size(); i++)
+                I *= src_dims[i];
+        } else if (layout == TopKLayoutType::topk_nspc) {
+            if (axis != 1) {
+                for (int i = 0; i < axis; i++) {
+                    if (i != 1)
+                        O *= src_dims[i];
+                }
+                I = src_dims[1];
+                for (size_t i = axis + 1; i < src_dims.size(); i++)
+                    if (i != 1)
+                        I *= src_dims[i];
+            } else {
+                for (int i = 0; i < src_dims_size; i++) {
+                    if (i != 1)
+                        O *= src_dims[i];
+                }
+            }
+        } else if (layout == TopKLayoutType::topk_blocked) {
+            N = src_dims[0];
+            ICB = div_up(src_dims[1], blk_size);
+            OCB = div_up(dst_dims[1], blk_size);
+            D = 1;
+            if (src_dims_size == 5)
+                D = src_dims[2];
+            H = src_dims[src_dims_size - 2];
+            W = src_dims[src_dims_size - 1];
+            if (axis == 0) { //topk on N
+                I = ICB * D * H * W * blk_size;
+            } else if (axis == 1) { //topk on C
+                if (algorithm != TopKAlgorithm::topk_bitonic_sort) {
+                    O = N * D * H * W;
+                } else {
+                    I = D * H * W;
+                }
+            } else if (src_dims_size == 5 && axis == 2) { //topk on D
+                O = N * ICB;
+                I = H * W * blk_size;
+            } else if (axis == src_dims_size - 2) { //topk on H
+                O = N * ICB * D;
+                I = W * blk_size;
+            } else if (axis == src_dims_size - 1) { //topk on W
+                O = N * ICB * D * H;
+                I = blk_size;
+            }
+        }
+
+        auto ceil_pow_2 = [&](size_t n, int &p, int &cnt, bool cmp_val) {
```

Comment:
I think this lambda name is not correctly describes its functionality.

Reply:
Fixed. Thanks Dmitry!